src/main/java/net/neoforged/neoforge/common/crafting/NBTIngredient.java

```diff
@@ -32,28 +37,46 @@ public class NBTIngredient extends Ingredient {
+                            ExtraCodecs.strictOptionalField(CraftingHelper.TAG_CODEC, AttachmentHolder.ATTACHMENTS_NBT_KEY).forGetter(NBTIngredient::getAttachmentNbt),
                             Codec.BOOL.optionalFieldOf("strict", false).forGetter(NBTIngredient::isStrict))
                     .apply(builder, NBTIngredient::new));
 
     public static final Codec<NBTIngredient> CODEC_NONEMPTY = RecordCodecBuilder.create(
             builder -> builder
                     .group(
                             NeoForgeExtraCodecs.singularOrPluralCodecNotEmpty(BuiltInRegistries.ITEM.byNameCodec(), "item").forGetter(NBTIngredient::getContainedItems),
-                            CraftingHelper.TAG_CODEC.fieldOf("tag").forGetter(NBTIngredient::getTag),
+                            ExtraCodecs.strictOptionalField(CraftingHelper.TAG_CODEC, "tag").forGetter(NBTIngredient::getOptionalTag),
+                            ExtraCodecs.strictOptionalField(CraftingHelper.TAG_CODEC, AttachmentHolder.ATTACHMENTS_NBT_KEY).forGetter(NBTIngredient::getAttachmentNbt),
                             Codec.BOOL.optionalFieldOf("strict", false).forGetter(NBTIngredient::isStrict))
                     .apply(builder, NBTIngredient::new));
 
     private final boolean strict;
 
+    @Deprecated(forRemoval = true, since = "1.20.4")
     protected NBTIngredient(Set<Item> items, CompoundTag tag, boolean strict) {
+        this(items, tag, null, strict);
+    }
+
+    @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+    private NBTIngredient(Set<Item> items, Optional<CompoundTag> tag, Optional<CompoundTag> attachmentTag, boolean strict) {
+        this(items, tag.orElse(null), attachmentTag.orElse(null), strict);
+    }
+
+    protected NBTIngredient(Set<Item> items, @Nullable CompoundTag tag, @Nullable CompoundTag attachmentTag, boolean strict) {
         super(items.stream().map(item -> {
-            ItemStack stack = new ItemStack(item, 1);
+            //Copy the attachment tag in case a modder mutates the tag they get passed when deserializing an attachment
+            ItemStack stack = new ItemStack(item, 1, attachmentTag == null ? null : attachmentTag.copy());
             // copy NBT to prevent the stack from modifying the original, as attachments or vanilla item durability will modify the tag
-            stack.setTag(tag.copy());
+            if (tag != null) {
+                stack.setTag(tag.copy());
+            }
             return new Ingredient.ItemValue(stack, strict ? ItemStack::matches : NBTIngredient::compareStacksWithNBT);
         }), NeoForgeMod.NBT_INGREDIENT_TYPE);
         Preconditions.checkArgument(!items.isEmpty(), "At least one item needs to be provided for a nbt matching ingredient");
+        if (!strict) {
+            Preconditions.checkArgument(tag != null || attachmentTag != null, "Either nbt or attachment nbt needs to be provided for a non-strict nbt matching ingredient");
+        }
```

Comment:
Would adding this check be a breaking change that should be deferred to 1.20.5?

Reply:
In 1.20.4 if there is no tag provided a NPE will occur

Reply:
ok, fair enough